spark/src/main/kotlin/com/adevinta/spark/res/AnnotatedStringResource.kt

```diff
@@ -0,0 +1,238 @@
+import androidx.compose.ui.text.SpanStyle
+import androidx.compose.ui.text.buildAnnotatedString
+import androidx.compose.ui.text.font.FontFamily
+import androidx.compose.ui.text.font.FontStyle
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.text.style.BaselineShift
+import androidx.compose.ui.text.style.TextDecoration
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.Density
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.em
+import androidx.core.text.HtmlCompat
+import com.adevinta.spark.PreviewTheme
+import com.adevinta.spark.R
+import com.adevinta.spark.SparkTheme
+import com.adevinta.spark.tokens.SparkColors
+import com.adevinta.spark.tokens.SparkTypography
+
+
+// FIXME: There is no official way to do this yet so we're waiting for
+//   https://issuetracker.google.com/issues/139320238 to be fixed
+@Composable
+@ReadOnlyComposable
+public fun resources(): Resources {
+    LocalConfiguration.current
+    return LocalContext.current.resources
+}
+
+public fun Spanned.toHtmlWithoutParagraphs(): String {
+    return HtmlCompat.toHtml(this, HtmlCompat.TO_HTML_PARAGRAPH_LINES_CONSECUTIVE)
+        .substringAfter("<p dir=\"ltr\">").substringBeforeLast("</p>")
+}
+
+public fun Resources.getText(@StringRes id: Int, vararg args: Any): CharSequence {
+    val escapedArgs = args.map {
+        if (it is Spanned) it.toHtmlWithoutParagraphs() else it
+    }.toTypedArray()
+    val resource = SpannedString(getText(id))
+    val htmlResource = resource.toHtmlWithoutParagraphs()
+    val formattedHtml = String.format(htmlResource, *escapedArgs)
+    return HtmlCompat.fromHtml(formattedHtml, HtmlCompat.FROM_HTML_MODE_LEGACY)
+}
+
+@Composable
+public fun annotatedStringResource(@StringRes id: Int, vararg formatArgs: Any): AnnotatedString {
+    val resources = resources()
+    val density = LocalDensity.current
+
+    val colors = SparkTheme.colors
+    val typography = SparkTheme.typography
+    return remember(id, formatArgs) {
+        val text = resources.getText(id, *formatArgs)
+        spannableStringToAnnotatedString(text, density, colors, typography)
+    }
+}
+
+@Composable
+public fun annotatedStringResource(@StringRes id: Int): AnnotatedString {
+    val resources = resources()
+    val density = LocalDensity.current
+
+    val colors = SparkTheme.colors
+    val typography = SparkTheme.typography
+    return remember(id) {
+        val text = resources.getText(id)
+        spannableStringToAnnotatedString(text, density, colors, typography)
+    }
+}
+
+private fun spannableStringToAnnotatedString(
+    text: CharSequence,
+    density: Density,
+    colors: SparkColors,
+    typography: SparkTypography,
+): AnnotatedString {
+    return if (text is Spanned) {
+        with(density) {
+            buildAnnotatedString {
+                append((text.toString()))
+                text.getSpans(0, text.length, Any::class.java).forEach {
```

Comment:
Nit: extract this inner loop into a `AnnotatedString.Builder` extension to reduce indentation and complexity.

Reply:
For example:


```kotlin

private fun spannableStringToAnnotatedString(
    text: CharSequence,
    density: Density,
    colors: SparkColors,
    typography: SparkTypography,
): AnnotatedString {
    if (text !is Spanned) return AnnotatedString(text.toString())
    return buildAnnotatedString {
        append(text.toString())
        text.getSpans(0, text.length, Any::class.java).forEach {
            buildWithSpan(it, text.getSpanStart(it), text.getSpanEnd(it))
        }
    }
}

private fun AnnotatedString.Builder.buildWithSpan(it: Any, start: Int, end: Int) = when (it) {
    is StyleSpan -> buildWithStyleSpan(it, start, end)
    is ...
}


private fun AnnotatedString.Builder.buildWithStyleSpan(it: StyleSpan, start: Int, end: Int) = when (it.style) {
    NORMAL -> FontWeight.Normal to FontStyle.Normal
    BOLD -> FontWeight.Bold to FontStyle.Normal
    ITALIC -> FontWeight.Normal to FontStyle.Italic
    BOLD_ITALIC -> FontWeight.Bold to FontStyle.Italic
    else -> null
}?.let { (fontWeight, fontStyle) ->
    addStyle(SpanStyle(fontWeight = fontWeight, fontStyle = fontStyle), start, end)
}
```